### DISCUSSION TITLE
Is Kotlin used in Xdrip and what about modules

### DISCUSSION THREAD

**Discussion by andyrozman** [NightscoutFoundation/xDrip#1995] [Q&A]
Hi !
Is Kotlin used in Xdrip or are we limited to Java... I am working on something that would like to contribute to Xdrip (in the future) and this is something that I would need to know? 

What about modules? Can I use modules or is that also no-no? 

What I have seen so far from code is that Kotlin is not used (but I am not sure about why), and modules are also not used (Xdrip has just app module). 

Thanks for answer.
Andy

P.S.: If there is a design document about Xdrip (about arhitecture), link would be welcome.

**Comment by tolot27** [comment]
There is an attempt to add Kotlin to xDrip (see #1401).

Also, there are three modules in xDrip: app, localeapi and wear. We plan to separate the identical code of app and wear into another module.

What is your background and what do you want to contribute to xDrip? Please discuss this with us, first. This increases the chance that your PR(s) get accepted. Also, it is important to commit nearly atomically and not to update libraries unless we agree on it.

Currently, there is no architecture documented. Please have a look at #1522, too and also on https://github.com/NightscoutFoundation/xDrip/projects/1. We try to fix the existing bugs with a higher priority than enhancements.

**Comment by andyrozman** [comment]
My Background... I am computer developer, I work mostly with Java (22 years), in last 4 years I started developing in Android, I am "pump-driver" developer on AndroidAPS project (Medtronic, Eros at beginning), in last year I worked with Kotlin, since a lot of stuff in AAPS was converted and rewriten into Kotlin. Long time ago I did contribute on Xdrip for very short while, so some of my files are still flowing arround.

I was thinkng about adding support for GlucoRx Aidex Cgms (I asked question if anybdy is working on that, but there was no answer). So my add would be only "driver" for this Cgms. I didn't plan any "core" changes to Xdrip, just support for another device. I have just started lloking at original Aidex code and will try to get it running and next few weeks, with my test application, and then I will see how far I get with that... If someone else want to do this, I can either colaborate with him/her or drop my "project" alltogether.

**Reply by tolot27** [reply]
Sounds great. We can create a [Gitter](https://gitter.im) room to discuss things further.
In general, we are interested in supporting more devices but cannot implement code which would be affected by DMCA or other copyright systems. But as you a developer of AAPS you know this problematic very well. :smile:

**Reply by andyrozman** [reply]
I am not sure about DMCA on this one (what is DMCA)... Like I said I am driver developer, so this licence things are well above my head... 
How did you handle this with Dexcom and other CGMS companies? 
Yes you can open gitter room, I still have access to Gitter... We can do discord too...
